### ISSUE TITLE
Add a "Copy code" button on code blocks

### ISSUE THREAD

**Issue by laurent22** [laurent22/joplin#2383]
When a code block is displayed, add a "Copy" button in the top-right hand corner. Clicking on it should copy the code to the clipboard. That button should only be displayed when the user moves the mouse over the code block.

## Platforms

- Desktop
- Mobile

* * *

![image](https://user-images.githubusercontent.com/1285584/73194694-812c1780-4124-11ea-989c-64dbb9d3b6ab.png)

**Comment by Runo-saduwa** [comment]
Please i will love to try to implement this, can this issue be assigned to me?, and if theres anyone more familier with the codebase, i will love to work together

**Comment by bedwardly-down** [comment]
@Runo-saduwa, I'm currently working on my own assignment here involving changes in the user interface and whatnot. I spent all night digging through a large chunk of the rendering and electron code and am getting fairly familiar with it, especially after the devs here started separating some of the dialog code. One very cool thing about the code here: most of the functions and class names are sensible and describe exactly what their overall function is for the most part, and those that don't make a whole lot of sense outright, it's pretty easy to figure out what they do by searching for where they're used in the code.

**Comment by laurent22** [comment]
@Runo-saduwa, at this point we only assign up to one issue per person.

**Comment by Runo-saduwa** [comment]
alright, thank

**Comment by bedwardly-down** [comment]
@laurent22 that is very reasonable. Thank you

**Comment by Runo-saduwa** [comment]
@laurent22 , i'm trying to give this a shot so if i understand you corretly, this is going to be implemented in this repo (joplin-renderer)[https://github.com/laurent22/joplin-renderer)]

**Comment by laurent22** [comment]
Yes indeed. Hmm now that the renderer is in a different repo it's not very easy to with with actually.

I'll see if I can think of something to make it easier, maybe by moving it back to this repo.

**Comment by Runo-saduwa** [comment]
Well, ive pulled the latest changes, i've been staring at the code for an hour now and i dont know where to start from, Lol............... can you kindly guide me through where to start from

**Comment by bedwardly-down** [comment]
@Runo-saduwa it looks like NoteRevisionViewer.jsx under ElectronClient/app/gui would be a good place to start. A lot of the editor rendering code is there. Then follow it where it takes you. The way each part of the app is linked is very sensible and logical.

**Comment by rheo-chiti** [comment]
Is this issue fixed for mobile yet,if not can I work upon this ?

**Comment by Runo-saduwa** [comment]
Sorry there’s already a pull request for this and it’s not a single
platform issue.

**Comment by HirenMadhu** [comment]
Is this issue still open?
If yes, then can I work on it?

**Comment by raghavg27** [comment]
Anyone working on this?

**Comment by tessus** [comment]
If there's no PR open, say you'll work on it and create a PR when ready.

We had so many people who said they'll work on something, but then nothing happened.
So, if you take an issue, please keep us apprised of the status.

**Comment by Rishgod** [comment]
> If there's no PR open, say you'll work on it and create a PR when ready.
> 
> We had so many people who said they'll work on something, but then nothing happened.
> So, if you take an issue, please keep us apprised of the status.

hey @tessus I would like to do this issue with the help of clipboardJS library, I will submit you the PR today

**Comment by flozender** [comment]
I would like to work on this, please.

**Comment by Rishgod** [comment]
> I would like to work on this, please.

Yes, @flozender  go ahead.

**Comment by Rishgod** [comment]
@laurent22 I have a question about the way to solve this, can we just not add the button HTML syntax here rather than writing the markdown plugin for it

![copy code](https://user-images.githubusercontent.com/54576074/78508293-43daaa80-77a3-11ea-8156-b417245b6707.jpg)

**Comment by kowalskidev** [comment]
I'm working on it 👍

**Comment by Rishgod** [comment]
> I'm working on it 👍

please don't I'm done with this PR

**Comment by wjrivera** [comment]
Was this feature implemented in [Pull Request 3024](https://github.com/laurent22/joplin/pull/3024) ?

**Comment by tessus** [comment]
No.

**Comment by wjrivera** [comment]
Is anyone assigned to this atm?

**Comment by tessus** [comment]
No, there were PRs at one point, but they were not merged. More info in #2426 and #3024

**Comment by ankitpal1029** [comment]
i'd like to work on this , now that joplin-renderer is archived where should i make changes

**Comment by Ash0311** [comment]
when will this be released ?

**Comment by tessus** [comment]
It won't, unless somebody picks up the PR a finishes it.

**Comment by ArenaGrenade** [comment]
Hey @tessus I am looking at the PRs made for this issue #2426 seems quite complete - except for the updates it needs on mobile. And #3024 seems to be a incomplete one and possibly stale for a long time.

What is the exact part of the feature that is yet to be done here? Is it that we need to make sure this works on mobile devices?

**Comment by laurent22** [comment]
This was incorrectly labelled as a "good first issue" but it's actually difficult to add this properly. And last year pull requests cannot be used a starting point as they weren't following the right approach.

**Comment by ArenaGrenade** [comment]
Ah! Sure.

**Comment by Ash0311** [comment]
Please add this soon as this is needed 1

**Comment by ArenaGrenade** [comment]
As of now what's the status of this issue? Is the button implemented anywhere - desktop / mobile?

**Comment by riekusr** [comment]
Would love to have this function as well

**Comment by RestlessRabbits** [comment]
Is there someone working on this?
I would also love to see this function :)

**Comment by kar0v** [comment]
I'm also missing this feature in joplin. Commenting for visibility.

**Comment by ckza** [comment]
Also commenting for visibility. Currently doing quite a bit of code and would love to have a nicer code block, with the copy button, to streamline my workflow. Otherwise I need to start using another note taking app and I would prefer to stick to Joplin.

**Comment by polytect** [comment]
Hi guys!

Can this feature be added to Joplin? It would attract even more users.

**Comment by Waldorf3** [comment]
Adding my support for this feature

**Comment by mntky** [comment]
I am using this plugin.
This plugin is working fine:)

https://github.com/LightAPIs/joplin-copy-code-blocks

**Comment by Coehill** [comment]
Any updates on this?

**Comment by tessus** [comment]
I doubt that this will be added to core. There are 3 plugins available that add this funtionality:

- [Copy Code Blocks](https://joplinapp.org/plugins/plugin/com.joplin.copy.codeBlocks/) v1.1.2
- [CodePerfect](https://joplinapp.org/plugins/plugin/jl15988.JoplinCodePerfectPlugin/) v1.0.2
- Code-Clipboard v1.0.0 (no longer maintained)

**Comment by Coehill** [comment]
> I doubt that this will be added to core. There are 3 plugins available that add this funtionality:
> 
>     * [Copy Code Blocks](https://joplinapp.org/plugins/plugin/com.joplin.copy.codeBlocks/) v1.1.2
> 
>     * [CodePerfect](https://joplinapp.org/plugins/plugin/jl15988.JoplinCodePerfectPlugin/) v1.0.2
> 
>     * Code-Clipboard v1.0.0 (no longer maintained)

Perhaps I'm misunderstanding, but if there are 3 plugins then doesn't that demonstrate a need for the functionality being in core?

**Comment by tessus** [comment]
Not necessarily. The plugin system was added to allow people (and the dev team) to add features to Joplin without having to change Joplin itself.

TBH many of the things that were added to core in the beginning could be handled as plugins now. I don't make any decisions so my reply is strictly a matter of personal opinion.

I don't see a reason to add anything that can be added via a plugin to the core code, but the school of thought might differ. If not, what is the appropriate criterium? The number of plugins that are available for a feature, the number of downloads of a certain plugin that implements a feature, rolling the dice? A combination or all of the above? I can't answer this with an absolute certainty. Can anyone?

**Comment by polytect** [comment]
I think this feature should have been included in the core when there were already two plugins available. Every programmer I know who has tried Joplin has had to resort to copying and pasting code while dragging the mouse as a first impression. The dragging method might highlight more text than the user intends. It's a standard functionality found in any website with Markdown codeblocks.

Not having this feature in the core may deter new users, as they need to be aware of the plugins option and take a chance on whether it will work. This can be a subtle turn-off.
It's often the small things that can attract or repel users in subtle ways, but that's a topic for another time.

I believe that Joplin's core should include the ability to copy codeblock contents, rather than relying on plugins. Just like having a light switch instead of a 
hotwire.


Don't copy me.
```
Copy me as quick as possible,
Does this have a copy option on GitHub?
```
Don't copy me.

**Comment by sourcecodemage** [comment]
> I think this feature should have been included in the core when there were already two plugins available. Every programmer I know who has tried Joplin has had to resort to copying and pasting code while dragging the mouse as a first impression. The dragging method might highlight more text than the user intends. It's a standard functionality found in any website with Markdown codeblocks.
> 
> Not having this feature in the core may deter new users, as they need to be aware of the plugins option and take a chance on whether it will work. This can be a subtle turn-off. It's often the small things that can attract or repel users in subtle ways, but that's a topic for another time.
> 
> I believe that Joplin's core should include the ability to copy codeblock contents, rather than relying on plugins. Just like having a light switch instead of a hotwire.
> 
> Don't copy me.
> 
> ```
> Copy me as quick as possible,
> Does this have a copy option on GitHub?
> ```
> 
> Don't copy me.

Great point. I'm considering migrating my Wiki type notes from Obsidian to Joplin or Bookstack . Joplin has more editor options, especially the "open in external app".  Bookstack comes with page navigation by section and the copy function for code blocks . I think those features will push me over to Bookstack.

**Comment by Waldorf3** [comment]
This request is now 5 years old. There is a plugin which kinda solves the problem, but it really should be a core functionality. Reading through this post it seems there are no lack of people who volunteer to work on it, yet nothing happens.

I'm sorry, but for all the nice things to say about joplin, it really seems a bit like abandonware at this time. Is there still active, and I mean, really active, not just having a look to see whats new every 6 months, development of this, or is it time to move to something else?

I came here from Evernotes, when they started to go full retard. It took me a bit of time to get used to it, but eventually it has grown on me, and I now use it as secondary memory. It disappoints me to see it being ignored.
